### PR TITLE
Support external regions

### LINKED ISSUE #162: Support user-defined external regions
Regarding external regions, we currently add `IMPEXP` and `MINRNW` to `ALL_REG`. Is there anything else we should be adding?

_Originally posted by @olejandro in https://github.com/etsap-TIMES/xl2times/issues/161#issuecomment-1893825697_

### PR DESCRIPTION
**Pull request by olejandro**
This PR closes #162

### CODE REVIEW

**Comment by olejandro** [comment]
@SamRWest would be interesting to try this on AusTIMES with single region. It drops data for regions that should be excluded early on, so should get to the breaking point faster. :)

**Comment by olejandro** [comment]
@siddharth-krishna the regression here (-1 record on Ireland) can be ignored. It is due to all uppercase regions by the tool.

**Comment by Antti-L** [comment]
I see there is a section **Remove external region columns from the all tables except tradelinks**.
Note that there are some parameters defined for external regions that must be preserved, at least `IRE_XBND`.

**Comment by olejandro** [comment]
Thanks, @Antti-L. Would be really nice to add a test case on this. Do you have anything available? It would also help ensuring that functionality won't break in the future versions. 

So far I've used Demo 7, removing one of the regions, for testing.

Also, do you have an overview of attributes, in addition to `IRE_XBND`, for which this is important?

File: xl2times/transforms.py [hunk 1]
```diff
@@ -866,15 +923,39 @@ def process_regions(
...
+    region_def["bookname"] = region_def[["bookname"]].ffill()
+    region_def = (
+        region_def.dropna(how="any")
+        .apply(lambda x: x.str.upper())
+        .drop_duplicates(ignore_index=True)
+    )
+    # Update model.all_regions
+    model.all_regions.update(region_def["region"].to_list())
+    # Determine model.internal_regions
+    booknames = set(region_def["bookname"].to_list())
```
Comment (siddharth-krishna) [inline xl2times/transforms.py:943]:
```suggestion
    model.all_regions.update(region_def["region"])
    # Determine model.internal_regions
    booknames = set(region_def["bookname"])
```
The `set` and `update` methods accept iterators, and df columns are Series, which are Iterable, so you don't need to convert to a list here.

File: xl2times/transforms.py [hunk 2]
```diff
@@ -866,15 +923,39 @@ def process_regions(
...
+    valid_booknames = {
+        b
+        for b in booknames
+        if any(re.match(rf"^VT_{b}_", file, re.IGNORECASE) for file in model.files)
+    }
     model.internal_regions.update(
-        utils.single_column(tables, datatypes.Tag.book_regions_map, "region")
+        region_def["region"][region_def["bookname"].isin(valid_booknames)].to_list()
```
Comment (siddharth-krishna) [inline xl2times/transforms.py:950]:
```suggestion
        region_def["region"][region_def["bookname"].isin(valid_booknames)]
```

File: xl2times/transforms.py [hunk 3]
```diff
@@ -111,6 +111,49 @@ def remove_comment_cols(table: datatypes.EmbeddedXlTable) -> datatypes.EmbeddedX
     return replace(table, dataframe=df)
 
 
+def remove_exreg_cols(
+    config: datatypes.Config,
+    tables: List[datatypes.EmbeddedXlTable],
+    model: datatypes.TimesModel,
+) -> List[datatypes.EmbeddedXlTable]:
+    """
+    Remove external region columns from the all tables except tradelinks.
```
Comment (siddharth-krishna) [inline xl2times/transforms.py:120]:
```suggestion
    Remove external region columns from all the tables except tradelinks.
```

**Review by siddharth-krishna — APPROVED** [review]
LGTM, just minor suggestions. Thanks (especially for adding explanatory comments!) 🚀

**Pull request merged**